### DISCUSSION TITLE
Hold on to filter when clicking a navigation item

### DISCUSSION THREAD

**Discussion by carlo-4ip** [netbox-community/netbox#12341] [Help Wanted!]
Hi all,

Is there a way I can hold on to a filter?

So if I filter on a tenant and go to a device, when I click on "Devices" in the navigation I lost my filter and see all devices from all tenants again. 

I would like to only lose my filter when I go to another tenant and click on Devices there, which would override the filter, or if I manually delete the current filter.

**Comment by candlerb** [comment]
What Netbox version are you using?  [v3.4.0](https://github.com/netbox-community/netbox/releases/v3.4.0) added the ability to save filters.

**Reply by carlo-4ip** [reply]
Hi, we're using v3.4.5, so yes we have the ability to save filters. But we have 50+ tenants, so to save all those filters would be a little time consuming.

That is why I would like to keep the current filter active, even when I open a Device and go back to the "Devices" table. Deleting the current filter is easy, but having to apply the filter each time we go back to the device list is a little frustrating.

**Reply by lurrkas** [reply]
This is a great idea that should be standard practise in all web applications that allow users to filter data. Applying a filter is supposed to save time and clicks. So having to click multiple times just to get to a filtered view defeats the purpose.

Watching the [Saved Filters](https://www.youtube.com/watch?v=oH1cGA4MUtY) video gave me hope (40s in) that navigating away would keep the filter active but no that was apparently not the point.

Bare minimum, a users saved filters should be accessible with a single click between the "Quick search" field and the "Results" tab under each section.

**Comment by ghaberek** [comment]
I've been looking for something like this as well. We have been using NetBox at my work for our own internal network but we're starting to use it for documenting existing client networks and designing new networks. It would be really helpful to navigate NetBox in "tenant mode." I imagine this as some sort of selector that would pre-filter the objects on any page (including the new dashboard) until the selector was changed or reset. This could also include a "None" option like in the Tenant filter selection, which would be helpful for finding objects that are not assigned to tenant but should be. I mocked up a rough example by inserting some HTML on the page:

|![image](https://user-images.githubusercontent.com/12944771/236021061-e41a1779-25f7-4984-ae48-2875a1f715d0.png)|![image](https://user-images.githubusercontent.com/12944771/236021102-d95d59e7-a775-42c0-a15f-53973b436d71.png)|
|:---:|:---:|
|Collapsed|Expanded|

**Reply by ghaberek** [reply]
Additional musings:
- I can see this causing confusion if a user forgets they selected (or accidentally selects) a tenant. Putting it front-and-center at the top of the page like I showed would help alleviate this, but it could also be shown only when there are two or more tenants.
- Correct me if I'm wrong, but I don't think this could be added with a plugin, as the [empty spacer column](../blob/develop/netbox/templates/base/layout.html#L50) in the navbar where I imagined putting this is part of the base page template, and the [documentation](https://docs.netbox.dev/en/stable/plugins/#limitations) states plugins cannot "override core templates."

**Reply by carlo-4ip** [reply]
Thank you for the addition to my question. 

I can see how this could work, I would like to add that a search field in such a filter drop down list, as we have 40 tenants.

**Reply by carlo-4ip** [reply]
@ghaberek, I don't see any replies from the developers.

Could you share your HTML code with me? Where did you insert that piece of coding?

**Reply by ghaberek** [reply]
Sorry I don't have the exact HTML I was testing with; it was a normal [dropdown](https://getbootstrap.com/docs/5.3/components/dropdowns/) component. I inserted it into [this div](https://github.com/netbox-community/netbox/blob/develop/netbox/templates/base/layout.html#L50) using the Chrome dev tools (right-click > Inspect) so it went away when I reloaded the page. I have not fiddled around with actually adding the feature in code.

**Reply by carlo-4ip** [reply]
Thank you Gregory!
That's to bad. 

I'm not sure I can do this myself, as there would need to a dynamic list of the tenants.


I hope the developers have an idea on how to create a filter that holds over different sections.

**Comment by carlo-4ip** [comment]
Dear developers,

Can you contribute to this challenge? 

The filtering of tenants is very important for us. Now I browse to the tenant and from there open the device list, then a filter is automatically applied. But when I have opened a device and want to return to the devices the tenant filter is gone.

I like the idea of Gregory, the ability to apply a tenant filter on a more global level.

I would really appreciate any help in this challenge.

**Comment by cmaz-bob** [comment]
Hi,

I'm very interested too by this idea. We have multiple site, and we'd like to be able to browse all the content just for 1 site, as you ask for tenants.

Thanks
Cyril

**Comment by Du10777** [comment]
I hit this same problem while doing infrastructure audits — I kept losing my tenant filter every time I moved between sections. Looking for a solution I came across this thread, which turned out to be a great source of ideas for what such a feature actually needs. So I built it:

**https://github.com/Du10777/netbox-scope-switcher**

It adds one compact selector to the top navbar. Pick your scope once and every list view stays scoped to it as you navigate — Devices → open a device → back to Devices keeps the filter. Exactly the behaviour asked for at the top of this thread.

Going through the specific points raised here:

- **@ghaberek** — your instinct about plugins not being able to touch the navbar was correct in 2023, but NetBox 4.1 added `PluginTemplateExtension.navbar()`, an officially supported hook for precisely this. No core template overrides needed. Your mockup is essentially what this ended up looking like.
- **The "None" option you suggested** — implemented. Each dimension has a `None` entry that maps to `?param=null`, for finding objects that should have a tenant assigned but don't.
- **Your concern about users forgetting a scope is active** — taken seriously. The control changes colour and shows the current selection directly in the navbar rather than hiding it behind a click, with a one-click clear. By default the plugin also issues a redirect that puts the filter params in the address bar, so NetBox's own filter panel shows them and URLs stay shareable.
- **@carlo-4ip's 50+ tenants / saved filters being too tedious** — no per-tenant setup of any kind. Nothing to create, nothing to maintain.
- **The search field you asked for** — every dimension has instant search in its picker, and multi-select, so 40+ tenants is fine.
- **Cyril's request for sites** — covered. Scope dimensions are configurable: tenant, tenant group, region, site group, site, location, tag, or any filter param you declare in `PLUGINS_CONFIG`. Dimensions cascade, so picking a tenant narrows the site list to sites where that tenant is actually present, and vice versa. Out of the box it's tenant-only, matching the original request.

Which dimensions apply on a given page is worked out by introspecting that view's filterset, so a prefixes list may scope by tenant and tag but not region, and the control dims when nothing applies. No models, no migrations, no external JS.

Requires NetBox 4.6 (CI also exercises 4.4 and 4.5). Feedback and issues very welcome — this thread shaped a lot of the feature set, so I'd be glad to hear whether it actually solves the problem for you.

**Reply by ghaberek** [reply]
*This is amazing!*

> @ghaberek — your instinct about plugins not being able to touch the navbar was correct in 2023, but NetBox 4.1 added `PluginTemplateExtension.navbar()`, an officially supported hook for precisely this. No core template overrides needed. Your mockup is essentially what this ended up looking like.

Yes, I've been digging into NetBox plugins lately and very recently discovered [`PluginTemplateExtension`](https://netboxlabs.com/docs/netbox/plugins/development/views/#extra-template-content). It's very powerful and I'm using it to inject additional info panels onto device pages. I had planned to revisit this once I completed the plugin I'm currently working on.

> **Your concern about users forgetting a scope is active** — taken seriously. The control changes colour and shows the current selection directly in the navbar rather than hiding it behind a click, with a one-click clear. By default the plugin also issues a redirect that puts the filter params in the address bar, so NetBox's own filter panel shows them and URLs stay shareable.

Excellent, that's exactly the way I would have done it as well. This may finally be the push I need for us to fully adopt NetBox at my work. We operate sort of like an MSP but for municipal transportation networks (and other things), so tenant separation is a *big* concern for us.

I had also experimented with combining the new [resource ownership](https://netboxlabs.com/docs/netbox/features/resource-ownership/) and [object-based permissions](https://netboxlabs.com/docs/netbox/administration/permissions/) features to segment users into tenant-specific groups that could view/edit unowned objects and objects owned by their groups, effectively hiding the other tenants. But the permissions settings started getting big and messy, and it seemed to create a lot of potential for human error,  so I scrapped the idea.

I'm so happy to have provided useful input for this. I'll keep you posted on how it works out. Kudos to you, sir.
